test/js_test.go

```diff
@@ -955,6 +955,250 @@ func TestJetStreamSubscribe(t *testing.T) {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
+		defer cancel()
+		res, err := sub.FetchChan(10, nats.Context(ctx))
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		if len(msgs) != 5 {
+			t.Fatalf("Expected %d messages; got: %d", 5, len(msgs))
+		}
+	})
+
+	t.Run("request expired", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 5; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		res, err := sub.FetchChan(10, nats.MaxWait(50*time.Millisecond))
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		if len(msgs) != 5 {
+			t.Fatalf("Expected %d messages; got: %d", 5, len(msgs))
+		}
+	})
+
+	t.Run("cancel context during fetch", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 5; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		res, err := sub.FetchChan(10, nats.Context(ctx))
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			cancel()
+		}()
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() == nil && !errors.Is(err, context.Canceled) {
+			t.Fatalf("Expected error: %s; got: %s", nats.ErrConsumerDeleted, err)
+		}
+		if len(msgs) != 5 {
+			t.Fatalf("Expected %d messages; got: %d", 5, len(msgs))
+		}
+	})
+
+	t.Run("remove durable consumer during fetch", func(t *testing.T) {
+		defer js.PurgeStream("TEST")
+		sub, err := js.PullSubscribe("foo", "cons")
+		if err != nil {
+			t.Fatalf("Unexpected error: %s", err)
+		}
+		for i := 0; i < 5; i++ {
+			if _, err := js.Publish("foo", []byte("msg")); err != nil {
+				t.Fatalf("Unexpected error: %s", err)
+			}
+		}
+		res, err := sub.FetchChan(10)
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			js.DeleteConsumer("TEST", "cons")
+		}()
+		msgs := make([]*nats.Msg, 0)
+		for msg := range res.Messages() {
+			msgs = append(msgs, msg)
+		}
+		if res.Error() == nil && !errors.Is(err, nats.ErrConsumerDeleted) {
+			t.Fatalf("Expected error: %s; got: %s", nats.ErrConsumerDeleted, err)
+		}
+		if len(msgs) != 5 {
+			t.Fatalf("Expected %d messages; got: %d", 5, len(msgs))
+		}
+	})
+
+	t.Run("validation errors", func(t *testing.T) {
```

Comment:
just in case, try to add a test where you have multiple fetch calls under different goroutines concurrently then a single select waiting for all the message channels

Reply:
done